prog/kube-utils/main.go

```diff
@@ -183,6 +191,51 @@ func reclaimPeer(weave weaveClient, cml *configMapAnnotations, peerName string,
 	return true, err
 }
 
+// regiesters with Kubernetes API server for node delete events. Node delete event handler
+// invokes reclaimRemovedPeers to remove it from IPAM so that IP space is reclaimed
+func registerForNodeUpdates(client *kubernetes.Clientset, stopCh <-chan struct{}, nodeName, peerName string) {
+	informerFactory := informers.NewSharedInformerFactory(client, 0)
+	nodeInformer := informerFactory.Core().V1().Nodes().Informer()
+	common.Log.Debugln("registering for updates for node delete events")
+	nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: func(obj interface{}) {
+			nodeObj, ok := obj.(*v1core.Node)
+			if !ok {
+				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+				if !ok {
+					common.Log.Fatalf("[kube-peers] Couldn't get node object from tombstone: %#v", obj)
+				}
+				nodeObj, ok = tombstone.Obj.(*v1core.Node)
+				if !ok {
+					common.Log.Fatalf("[kube-peers] Tombstone contained object that is not a Node: %#v", obj)
+				}
+			}
+			common.Log.Debugln("[kube-peers] Nodes deleted:", nodeObj.Name)
+			config, err := rest.InClusterConfig()
```

Comment:
this is a bit weird - surely we already have a connection to Kubernetes otherwise we wouldn't be in a callback